### ISSUE TITLE
习题4-1

### ISSUE THREAD

**Issue by fecet** [nndl/solutions#9]
![4-1](https://user-images.githubusercontent.com/41792945/77856584-b63f0e00-722a-11ea-9a71-81458967ee7e.png)

**Comment by douzujun** [comment]
非零中心化的输出会使得其后一层的神经元的输入发生 **偏置偏移**，并进一步使得梯度下降的收敛速度变慢。

**Comment by progress612** [comment]
此外，接近 0 的输入在 sigmoid 型函数上的导数较大，梯度下降速度较快

**Comment by zzgnb** [comment]
![image](https://github.com/nndl/solutions/assets/51527848/6ce1657b-d620-4504-a77f-2e577ff1429d)
在更新式w=w+yx中，输入x恒大于0即所有维x>0，共用一个y，则所有维yx同号，w中各维的梯度更新方向都为正或都为负；
并不是说梯度更新的符号不变，而是所有维度的梯度更新符号一致，这就减少了梯度下降可选择的方向；
例如在图中的二维空间中，由于梯度全为正或负，因此梯度更新仅能向右上（w梯度>0）或左下（w梯度<0）更新，而若最优解在当前位置的右下（w0梯度>0，w1梯度<0）时，则只能通过振荡的方法逼近，影响梯度下降速度

**Comment by mJyo** [comment]
```python
import torch
import torch.nn as nn
import torch.optim as optim
import matplotlib.pyplot as plt

# 定义简单的神经网络
class SimpleNet(nn.Module):
    def __init__(self):
        super(SimpleNet, self).__init__()
        self.fc = nn.Linear(1, 1)
        self.sigmoid = nn.Sigmoid()
    
    def forward(self, x):
        x = self.fc(x)
        return self.sigmoid(x)

# 生成数据：恒大于0的数据
torch.manual_seed(0)
x_positive = torch.rand(100, 1) * 10  # 0到10之间的正数
y = 2 * x_positive + 1  # 简单的线性关系

# 生成数据：零均值化的数据
x_zero_mean = (x_positive - x_positive.mean()) / x_positive.std()

# 定义损失函数和优化器
criterion = nn.MSELoss()

# 训练模型函数
def train_model(x_data, num_epochs=100):
    model = SimpleNet()
    optimizer = optim.SGD(model.parameters(), lr=0.01)
    loss_history = []
    
    for epoch in range(num_epochs):
        model.train()
        
        # 前向传播
        outputs = model(x_data)
        loss = criterion(outputs, y)
        
        # 反向传播和优化
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        loss_history.append(loss.item())
    
    return loss_history

# 训练模型并获取损失曲线
loss_positive = train_model(x_positive)
loss_zero_mean = train_model(x_zero_mean)

# 绘制损失曲线
plt.plot(loss_positive, label='Positive Input (x > 0)')
plt.plot(loss_zero_mean, label='Zero-mean Input')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.title('Loss Curve Comparison')
plt.legend()
plt.show()
```
